### PR TITLE
feat(dh): frontend to backend-for-frontend to domain communication

### PR DESCRIPTION
**Pull request by LayZeeDK**
### Frontend
- Use local backend-for-frontend for local development
- Use Development backend-for-frontend when deployed

### Backend-for-frontend
- Add lax CORS policy for the Development configuration to allow requests from the frontend
- Use metering point testing environment for local development as well as unit and integration tests
- Output static OpenAPI 3 spec on build
- Add Nx `build` target
- Add Nx `build-client` target to generate DataHub API HTTP client and related types
- Add Nx `ci-configuration` target to copy `appsettings.Development.json` configuration to API folder
- Add Nx `serve` target for local development and tests

### Test staging
- Configure Jest to allow using a local backend-for-frontend for unit, integration, and end-to-end tests

## References

- https://github.com/Energinet-DataHub/greenforce-frontend/issues/122
- https://github.com/Energinet-DataHub/greenforce-frontend/issues/134

## PLEASE NOTE

Before merging a frontend pull request, be sure that all checks have passed. Currently, jobs done by a Git-bot (like formatting and adding licenses) don't trigger a re-run of all the other checks. The "merge button" can therefore give you a false positive signal.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -12,6 +12,8 @@
     "postinstall": "npm run postinstall:nx && npm run postinstall:ngcc",
     "postinstall:ngcc": "npm run ngcc",
     "postinstall:nx": "node ./decorate-angular-cli.js",
+    "check:dotnet": "dotnet --version || echo ❌ dotnet CLI must be installed",
+    "check:java": "java -version || echo ❌ Java must be installed",
```
Comment (dzhavat) [inline package.json:16]: question: What is Java used for?

**Comment by dzhavat** [comment]
Looks quite good. There are some minor stuff.

File: apps/dh/api-dh/project.json [hunk 2]
```diff
@@ -3,6 +3,81 @@
...
+          "commands": [
+            "npm run check:dotnet",
+            "npx ts-node tools/dh/api-dh/ci/start-api-dh.ts"
+          ]
+        }
+      },
+      "executor": "@nrwl/workspace:run-commands",
+      "options": {
+        "commands": [
+          "npm run check:dotnet",
```
[the remark this answers is not in the document]
Comment (LayZeeDK) [reply apps/dh/api-dh/project.json:62]: Yarn spawns another process which sometimes removes output when running as part of other CI commands. `npm run` doesn't have this issue.

File: package.json (hunk 1 above)
Comment (LayZeeDK) [reply package.json:16]: Java is used for the OpenAPI-to-Angular HTTP client generation.

File: apps/dh/api-dh/project.json [hunk 3]
```diff
@@ -3,6 +3,81 @@
...
+        "parallel": false
+      }
+    },
+    "serve": {
+      "configurations": {
+        "ci": {
+          "commands": [
+            "npm run check:dotnet",
+            "npx ts-node tools/dh/api-dh/ci/start-api-dh.ts"
+          ]
```
[the remark this answers is not in the document]
Comment (LayZeeDK) [reply apps/dh/api-dh/project.json:56]: The targets are mostly used by CI.

The `serve` command should be used by us. We need to spin up a DataHub API (BFF) locally to run tests and perform local development. Where is a good place to document this?

File: libs/ett/shared/util-browser/jest.config.js [hunk 4]
```diff
@@ -33,4 +33,5 @@ module.exports = {
     'jest-preset-angular/build/serializers/ng-snapshot',
     'jest-preset-angular/build/serializers/html-comment',
   ],
+  testURL: 'http://localhost',
```
[the remark this answers is not in the document]
Comment (LayZeeDK) [reply libs/ett/shared/util-browser/jest.config.js:36]: The default Jest configuration has `https://localhost:5001` to allow requests to a local DataHub API (BFF) instance.

I've adjusted the tests in this project to take this into account.

File: apps/dh/api-dh/project.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (dzhavat) [reply apps/dh/api-dh/project.json:56]: > Where is a good place to document this?

What if we had a command similar to `check:dotnet` that we can run at the start of each `commands` array to put a message in the console making the developer aware of them? Something like

```json
"nx:customCommand": "Executing custom commands in project.json file",
```

[the remark this answers is not in the document]
Comment (LayZeeDK) [reply apps/dh/api-dh/project.json:56]: They're alread used as `nx run <project>:<target>[:<configuration>]` or `nx <target> <project>[:<configuration>]`

**Review by dzhavat — APPROVED** [review]

**Pull request merged**